### PR TITLE
Add ed25519 to x25519 to nsec.experimental

### PR DESCRIPTION
**Pull request by sigaloid**
Still working out how to take Key instances.

### CODE REVIEW

File: src/Experimental/Ed25519ToX25519.cs [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+            var xPublicKey = new byte[32];
+
+            fixed (byte* x25519_sk = xPublicKey)
+            fixed (byte* ed25519_sk = publicKey.Export(KeyBlobFormat.RawPublicKey))
+            {
+                int error = crypto_sign_ed25519_pk_to_curve25519(x25519_sk, ed25519_sk);
+                Debug.Assert(error == 0);
+            }
+
+            return PublicKey.Import(KeyAgreementAlgorithm.X25519, xPublicKey, KeyBlobFormat.RawPublicKey);
```
Comment (sigaloid) [inline src/Experimental/Ed25519ToX25519.cs:32]: Shouldn't be exporting/importing here

File: src/Experimental/Ed25519ToX25519.cs [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+            }
+
+            return Key.Import(KeyAgreementAlgorithm.X25519, xPrivateKey, KeyBlobFormat.RawPrivateKey, new KeyCreationParameters { ExportPolicy = KeyExportPolicies.AllowPlaintextExport});
+        }
+        public unsafe static PublicKey SignEd25519PublicToCurve25519(PublicKey publicKey)
+        {
+            var xPublicKey = new byte[32];
+
+            fixed (byte* x25519_sk = xPublicKey)
+            fixed (byte* ed25519_sk = publicKey.Export(KeyBlobFormat.RawPublicKey))
```
Comment (sigaloid) [inline src/Experimental/Ed25519ToX25519.cs:26]: Shouldn't be exporting/importing here

File: src/Experimental/Ed25519ToX25519.cs [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+            var xPrivateKey = new byte[32];
+
+            fixed (byte* x25519_sk = xPrivateKey)
+            fixed (byte* ed25519_sk = privateKey.Export(KeyBlobFormat.RawPrivateKey))
+            {
+                int error = crypto_sign_ed25519_sk_to_curve25519(x25519_sk, ed25519_sk);
+                Debug.Assert(error == 0);
+            }
+
+            return Key.Import(KeyAgreementAlgorithm.X25519, xPrivateKey, KeyBlobFormat.RawPrivateKey, new KeyCreationParameters { ExportPolicy = KeyExportPolicies.AllowPlaintextExport});
```
Comment (sigaloid) [inline src/Experimental/Ed25519ToX25519.cs:19]: Shouldn't be exporting/importing here

File: src/Experimental/Ed25519ToX25519.cs [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+namespace NSec.Cryptography
+{
+    public static class Ed25519ToX25519
+    {
+        public unsafe static Key SignEd25519SecretToCurve25519(Key privateKey)
+        {
+            var xPrivateKey = new byte[32];
+
+            fixed (byte* x25519_sk = xPrivateKey)
+            fixed (byte* ed25519_sk = privateKey.Export(KeyBlobFormat.RawPrivateKey))
```
Comment (sigaloid) [inline src/Experimental/Ed25519ToX25519.cs:13]: Shouldn't be exporting/importing here

**Review by sigaloid — COMMENTED** [review]
Shouldn't be exporting/importing here. 😢

File: tests/Algorithms/Ed25519ToX25519Test.cs [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+    public class Ed25519ToX25519Test
+    {
+        [Fact]
+        public void TestEd25519_to_X25519()
+        {
+            var ed25519 = Key.Create(SignatureAlgorithm.Ed25519,
+                new KeyCreationParameters() {ExportPolicy = KeyExportPolicies.AllowPlaintextExport});
+            var x25519pub = Ed25519ToX25519.SignEd25519PublicToCurve25519(ed25519.PublicKey);
+            var x25519priv = Ed25519ToX25519.SignEd25519SecretToCurve25519(ed25519);
+            Assert.Equal(x25519priv.PublicKey, x25519pub);
```
Comment (sigaloid) [inline tests/Algorithms/Ed25519ToX25519Test.cs:17]: Does this check that the key here is actually correct (ie, key generated corresponds with the ed25519 key) or that the key is simply valid (ie, public key returned corresponds with the private key returned)?

**Comment by sigaloid** [comment]
Forgot to mention this is from #32, thanks @p3root

File: tests/Algorithms/Ed25519ToX25519Test.cs (hunk 5 above)
Comment (ektrah) [reply tests/Algorithms/Ed25519ToX25519Test.cs:17]: There are basically two paths to get a converted public key:

```
Ed25519      SignEd25519SecretToCurve25519      X25519
private   ----------------------------------->  private
  key                                             key

   |                                              |
   | .PublicKey                        .PublicKey |
   |                                              |
   v                                              v

Ed25519      SignEd25519PublicToCurve25519      X25519
public    ----------------------------------->  public
  key                                             key
```

This test just checks that the converted public keys for both of the paths are actually the same (which is a good thing to test!).

It doesn't test whether the conversion was correct, though. So the next step would be to add some tests with verified test vectors to ensure that everything is right.

**Comment by ektrah** [comment]
Thanks a lot for your pull request! (We're moving slowly, but we're moving 😀)

I've made a few changes in 27572953586ead37f53f0523ff654cd804083894 to avoid exporting/importing. Let me know if that works for you.

File: tests/Algorithms/Ed25519ToX25519Test.cs (hunk 5 above)
Comment (sigaloid) [reply tests/Algorithms/Ed25519ToX25519Test.cs:17]: Thanks!

**Comment by sigaloid** [comment]
:+1:

**Pull request merged**